### PR TITLE
Provide API key when uploading dSYMs

### PR DESCRIPTION
**Pull request by CodeHex**
## Goal

When a dSYM is uploaded, the API key should be included when making the call to the upload server. For all plugins and tools, we should try to do this by default if possible (by reading the Info.plist or via an environment variable).

### Added
 - The API key parameter has been added to the bugsnag-dsym binary
 - Fastlane plugin supports API key parametrer, which is read from the Info.plist by default
 - Fastlane plugin supports config file parameter, allowing you to provide an alternate location for the Info.plist

### Changed
 - Fastlane plugin tests use default options rather than no options.

## Tests

Tests were added for the new bugsnag-dsym binary and Fastlane plugin functionality
Inital integration test run with the Fastlane plugin
- [x] Run final integration test with Fastlane plugin
- [x] Run final integration test with bugsnag-dsym binary

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [x] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [x] Consistency across platforms for structures or concepts added or modified
- [x] Consistency between the changeset and the goal stated above
- [x] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [x] Usage friction - is the proposed change in usage cumbersome or complicated?
- [x] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [x] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: bin/bugsnag-dsym-upload [hunk 1]
```diff
@@ -71,6 +72,10 @@ while [[ $# -gt 0 ]]; do
             upload_server=$2
             shift
             shift;;
+        --api-key)
+            api_key=$2
```
Comment (mikeewhite) [inline bin/bugsnag-dsym-upload]: Does this need an `unset` entry above?

File: man/bugsnag-dsym-upload.pod [hunk 2]
```diff
@@ -53,6 +54,11 @@ Optional, only required when uploading a ZIP archive.
 
 Display the help documentation for this command
 
+=item --api-key API_KEY
+
+The API key associated with the project. Informs Bugsnag which project this dSYM should
+be applied to.
+
```
Comment (mikeewhite) [inline man/bugsnag-dsym-upload.pod]: Should we also add an entry for project root?

File: tools/fastlane-plugin/spec/bugsnag_upload_dsym_action_spec.rb [hunk 3]
```diff
@@ -103,17 +112,50 @@
...
+    
+    it 'allows config file to overwrite parameter' do
+      expect(Kernel).to receive(:system).with(Action::UPLOAD_SCRIPT_PATH,
+                                              "--silent",
+                                              "--api-key", "project-key",
+                                              "--project-root", File.join(FIXTURE_PATH, 'ios_proj'),
+                                              FIXTURE_PATH).and_return(true)
+      Dir.chdir(File.join(FIXTURE_PATH, 'ios_proj')) do
+        Action.run(load_default_opts.merge({dsym_path: FIXTURE_PATH, api_key: "ignored:", config_file: File.join('Project', 'Info.plist')}))
+      end
```
Comment (mikeewhite) [inline tools/fastlane-plugin/spec/bugsnag_upload_dsym_action_spec.rb]: It might be worth adding a test where `api-key` is without `project-root`

File: bin/bugsnag-dsym-upload [hunk 4]
```diff
@@ -131,12 +136,23 @@ for dsym in $dsym_dir/*.dSYM; do
         uuid=$(dwarfdump -u $file 2>/dev/null)
         if [[ $uuid == UUID* ]]; then
             log Uploading $uuid
+
+            # Attach the api key and project root parameters if they have been provided
+            arg=""
```
Comment (mikeewhite) [inline bin/bugsnag-dsym-upload]: I think this should be `args`.

File: tools/fastlane-plugin/lib/fastlane/plugin/bugsnag/actions/upload_symbols_to_bugsnag.rb [hunk 5]
```diff
@@ -59,7 +71,15 @@ def self.available_options
...
         end
+
+        options = {}
+        options = options_from_info_plist(default_info_plist_path) if file_path = default_info_plist_path
         [
+          FastlaneCore::ConfigItem.new(key: :api_key,
+                                       env_name: "BUGSNAG_API_KEY",
+                                       description: "Bugsnag API Key",
+                                       optional: true,
+                                       default_value: options[:apiKey]),
```
Comment (Cawllec) [inline tools/fastlane-plugin/lib/fastlane/plugin/bugsnag/actions/upload_symbols_to_bugsnag.rb]: Would it be a good idea to add some validation here?  How does the upload server react if an invalid API key is used?

File: README.md [hunk 6]
```diff
@@ -50,6 +50,8 @@ end
 
 Common options:
 
+* `api_key`: The API key associated with the project. Informs Bugsnag which project
```
Comment (Cawllec) [inline README.md]: We should add the `config_file` options here as well.

File: features/dsym_upload.feature [hunk 7]
```diff
@@ -16,3 +16,18 @@ Feature: Uploading dSYMs to Bugsnag
...
+        And the part "dsym" for request 0 is not null
+        And the part "apiKey" for request 0 equals "1234567890ABCDEF"
+        And the payload field "projectRoot" is null for request 0
+
+    Scenario: Uploading dSYMs with a project root and API key
+        When I upload dSYMS with options "--project-root /Users/jenkins/build/app/ --api-key 1234567890ABCDEF features/fixtures"
+        Then I should receive a request
+        And the part "dsym" for request 0 is not null
+        And the part "projectRoot" for request 0 equals "/Users/jenkins/build/app/"
+        And the part "apiKey" for request 0 equals "1234567890ABCDEF"
```
Comment (Cawllec) [inline features/dsym_upload.feature]: It would be good to have a scenario with conflicting api-key options and config-file options, to ensure config-file takes priority.

File: tools/fastlane-plugin/lib/fastlane/plugin/bugsnag/actions/upload_symbols_to_bugsnag.rb [hunk 8]
```diff
@@ -6,9 +6,21 @@ class UploadSymbolsToBugsnagAction < Action
...
+        options = {}
+        options = options_from_info_plist(params[:config_file]) if params[:config_file]
+
+        if (params[:config_file] == default_info_plist_path)
+          # Load custom API key if config file has not been overridden
+          options[:apiKey] = params[:api_key] unless params[:api_key].nil?
+        else
+          # Print which file is populating version and API key information since the value has been
+          # overridden
+          UI.message("Loading API key from #{params[:config_file]}")
```
Comment (Cawllec) [inline tools/fastlane-plugin/lib/fastlane/plugin/bugsnag/actions/upload_symbols_to_bugsnag.rb]: What happens if the `config_file` option is present but the `BugsnagApiKey` is not inside it?

File: features/dsym_upload.feature (hunk 7 above)
Comment (Cawllec) [reply features/dsym_upload.feature]: I did see the test for this below, but I think it'd be good to have it in maze as well.

File: tools/fastlane-plugin/lib/fastlane/plugin/bugsnag/actions/upload_symbols_to_bugsnag.rb (hunk 8 above)
Comment (Cawllec) [reply tools/fastlane-plugin/lib/fastlane/plugin/bugsnag/actions/upload_symbols_to_bugsnag.rb]: Tested this - if you have a `config_file` parameter with no API key, the `api_key` parameter will be ignored.  We should change this to use the `api_key` parameter if not api key can be extracted from `info.plist`.

File: tools/fastlane-plugin/lib/fastlane/plugin/bugsnag/actions/upload_symbols_to_bugsnag.rb (hunk 5 above)
Comment (CodeHex) [reply tools/fastlane-plugin/lib/fastlane/plugin/bugsnag/actions/upload_symbols_to_bugsnag.rb]: I've added some default validation, but this doesn't always trigger depending on where the api key is provided. The upload server responds with invalid apikey and fails the upload, so hopefully this should be enough to indicate whats gone wrong

File: features/dsym_upload.feature (hunk 7 above)
Comment (CodeHex) [reply features/dsym_upload.feature]: I've added some more tests with a few variations. Need to clean up the test fastlane file as the plugin reacts differently depending on if variables are passed explicitly or read from the environment

File: tools/fastlane-plugin/lib/fastlane/plugin/bugsnag/actions/upload_symbols_to_bugsnag.rb (hunk 8 above)
Comment (CodeHex) [reply tools/fastlane-plugin/lib/fastlane/plugin/bugsnag/actions/upload_symbols_to_bugsnag.rb]: Added a test for this, will now use the parameter if there is nothing in the config file.

**Comment by snmaynard** [comment]
Merge this at the same time https://github.com/bugsnag/docs.bugsnag.com/pull/809
